js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,575 @@
+	}
+
+	if method[0] != '/' {
+		method = "/" + method
+	}
+	md := c.mds[method]
+
+	if md == nil {
+		return nil, fmt.Errorf("method %q not found in file descriptors", method)
+	}
+
+	tags := state.CloneTags()
+	timeout := 60 * time.Second
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.New(nil))
+	for k, v := range params {
+		switch k {
+		case "headers":
+			rawHeaders, ok := v.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("headers must be an object with key-value pairs")
+			}
+			for hk, kv := range rawHeaders {
+				// TODO(rogchap): Should we manage a string slice?
+				strVal, ok := kv.(string)
+				if !ok {
+					return nil, fmt.Errorf("header %q value must be a string", hk)
+				}
+				ctx = metadata.AppendToOutgoingContext(ctx, hk, strVal)
+			}
+		case "tags":
+			rawTags, ok := v.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("tags must be an object with key-value pairs")
+			}
+			for tk, tv := range rawTags {
+				strVal, ok := tv.(string)
+				if !ok {
+					return nil, fmt.Errorf("tag %q value must be a string", tk)
+				}
+				tags[tk] = strVal
+			}
+		case "timeout":
+			var err error
+			switch t := v.(type) {
+			case string:
+				if timeout, err = time.ParseDuration(t); err != nil {
+					return nil, fmt.Errorf("unable to parse %q: %v", k, err)
+				}
+			case int64:
+				timeout = time.Duration(float64(t)) * time.Millisecond
+			case float64:
+				timeout = time.Duration(t) * time.Millisecond
+			default:
+				return nil, fmt.Errorf("unable to use type %T as a timeout value", v)
+			}
+		default:
+			return nil, fmt.Errorf("unknown param: %q", k)
+		}
+	}
+	if state.Options.SystemTags.Has(stats.TagURL) {
+		tags["url"] = fmt.Sprintf("%s%s", c.conn.Target(), method)
+	}
+
+	parts := strings.Split(method[1:], "/")
+	if state.Options.SystemTags.Has(stats.TagService) {
+		tags["service"] = parts[0]
+	}
+	if state.Options.SystemTags.Has(stats.TagMethod) {
+		tags["method"] = parts[1]
+	}
+
+	// Only set the name system tag if the user didn't explicitly set it beforehand
+	if _, ok := tags["name"]; !ok && state.Options.SystemTags.Has(stats.TagName) {
+		tags["name"] = method
+	}
+
+	ctx = withTags(ctx, tags)
+
+	reqdm := dynamicpb.NewMessage(md.Input())
+	{
+		b, err := req.ToObject(rt).MarshalJSON()
+		if err != nil {
+			return nil, fmt.Errorf("unable to serialise request object: %v", err)
+		}
+		if err := protojson.Unmarshal(b, reqdm); err != nil {
+			return nil, fmt.Errorf("unable to serialise request object to protocol buffer: %v", err)
+		}
+	}
+
+	reqCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	resp := dynamicpb.NewMessage(md.Output())
+	header, trailer := metadata.New(nil), metadata.New(nil)
+	err := c.conn.Invoke(reqCtx, method, reqdm, resp, grpc.Header(&header), grpc.Trailer(&trailer))
+
+	var response Response
+	response.Headers = header
+	response.Trailers = trailer
+
+	marshaler := protojson.MarshalOptions{EmitUnpopulated: true}
+
```

Comment:
I do think it will be more readable if this is moved before the Invoke so taht `if err != nil` can be directly after the Invoke. 
Also making the ctx should probably be directly before the Invoke and you can skip the header/trailer varaible and directly use response.Headers/response.Trailers as far as I can see

Reply:
If I directly use `response.Headers`/`response.Trailers` this results in issues with casting, as one is `map[string][]string` and the other is `metadata.MD` which is an alias to `map[string][]string`. If I change the `Response` struct to use `metadata.MD` then this will not serialize correctly via goja (unmarshals to an empty object `{}`)